**Name:-chandan birendra pandit            
Position :-Machine Learning Intern                       
ref:- CTI/A1/C76432**

Task: Cuisine Classification


*   Objective: Develop a machine learning model to
 classify restaurants based on their cuisines.





# Data Preprocess

Preprocess the dataset by handling missing values,
encoding categorical variables, and splitting the data
into training and testing sets.

Importing necessary libraries


In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv('/content/Dataset .csv')

In [3]:
df.head(2)

Restaurant ID   Restaurant Name  Country Code         City  \
0        6317637  Le Petit Souffle           162  Makati City   
1        6304287  Izakaya Kikufuji           162  Makati City   

                                             Address  \
0  Third Floor, Century City Mall, Kalayaan Avenu...   
1  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   

                                     Locality  \
0   Century City Mall, Poblacion, Makati City   
1  Little Tokyo, Legaspi Village, Makati City   

                                    Locality Verbose   Longitude   Latitude  \
0  Century City Mall, Poblacion, Makati City, Mak...  121.027535  14.565443   
1  Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101  14.553708   

                     Cuisines  ...          Currency Has Table booking  \
0  French, Japanese, Desserts  ...  Botswana Pula(P)               Yes   
1                    Japanese  ...  Botswana Pula(P)               Yes   

  Has Online delivery Is delivering now Switch to order menu Price range  \
0                  No                No                   No           3   
1                  No                No                   No           3   

   Aggregate rating  Rating color Rating text Votes  
0               4.8    Dark Green   Excellent   314  
1               4.5    Dark Green   Excellent   591  

[2 rows x 21 columns]

In [4]:
df.shape

(9551, 21)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9551 entries, 0 to 9550
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Restaurant ID         9551 non-null   int64  
 1   Restaurant Name       9551 non-null   object 
 2   Country Code          9551 non-null   int64  
 3   City                  9551 non-null   object 
 4   Address               9551 non-null   object 
 5   Locality              9551 non-null   object 
 6   Locality Verbose      9551 non-null   object 
 7   Longitude             9551 non-null   float64
 8   Latitude              9551 non-null   float64
 9   Cuisines              9542 non-null   object 
 10  Average Cost for two  9551 non-null   int64  
 11  Currency              9551 non-null   object 
 12  Has Table booking     9551 non-null   object 
 13  Has Online delivery   9551 non-null   object 
 14  Is delivering now     9551 non-null   object 
 15  Switch to order menu 

In [6]:
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

In [7]:
df.columns

Index(['Restaurant ID', 'Restaurant Name', 'Country Code', 'City', 'Address',
       'Locality', 'Locality Verbose', 'Longitude', 'Latitude', 'Cuisines',
       'Average Cost for two', 'Currency', 'Has Table booking',
       'Has Online delivery', 'Is delivering now', 'Switch to order menu',
       'Price range', 'Aggregate rating', 'Rating color', 'Rating text',
       'Votes'],
      dtype='object')

In [8]:
df['Restaurant ID'].unique()

array([6317637, 6304287, 6300002, ..., 5915807, 5916112, 5927402])

In [9]:
df['Restaurant Name'].unique()

array(['Le Petit Souffle', 'Izakaya Kikufuji', 'Heat - Edsa Shangri-La',
       ..., 'Huqqa', 'A���k Kahve', "Walter's Coffee Roastery"],
      dtype=object)

In [10]:
df['Country Code'].unique()

array([162,  30, 216,  14,  37, 184, 214,   1,  94, 148, 215, 166, 189,
       191, 208])

In [11]:
df['Country Code'].value_counts()

Country Code
1      8652
216     434
215      80
30       60
214      60
189      60
148      40
208      34
14       24
162      22
94       21
184      20
166      20
191      20
37        4
Name: count, dtype: int64

In [12]:
df['City'].unique()

array(['Makati City', 'Mandaluyong City', 'Pasay City', 'Pasig City',
       'Quezon City', 'San Juan City', 'Santa Rosa', 'Tagaytay City',
       'Taguig City', 'Bras�_lia', 'Rio de Janeiro', 'S��o Paulo',
       'Albany', 'Armidale', 'Athens', 'Augusta', 'Balingup',
       'Beechworth', 'Boise', 'Cedar Rapids/Iowa City', 'Chatham-Kent',
       'Clatskanie', 'Cochrane', 'Columbus', 'Consort', 'Dalton',
       'Davenport', 'Des Moines', 'Dicky Beach', 'Dubuque',
       'East Ballina', 'Fernley', 'Flaxton', 'Forrest', 'Gainesville',
       'Hepburn Springs', 'Huskisson', 'Inverloch', 'Lakes Entrance',
       'Lakeview', 'Lincoln', 'Lorn', 'Macedon', 'Macon', 'Mayfield',
       'Mc Millan', 'Middleton Beach', 'Miller', 'Monroe', 'Montville',
       'Ojo Caliente', 'Orlando', 'Palm Cove', 'Paynesville', 'Penola',
       'Pensacola', 'Phillip Island', 'Pocatello', 'Potrero', 'Princeton',
       'Rest of Hawaii', 'Savannah', 'Singapore', 'Sioux City',
       'Tampa Bay', 'Tanunda', 'Trentha

In [13]:
df['City'].value_counts()

City
New Delhi           5473
Gurgaon             1118
Noida               1080
Faridabad            251
Ghaziabad             25
                    ... 
Panchkula              1
Mc Millan              1
Mayfield               1
Macedon                1
Vineland Station       1
Name: count, Length: 141, dtype: int64

In [14]:
df['Address'].unique()

array(['Third Floor, Century City Mall, Kalayaan Avenue, Poblacion, Makati City',
       'Little Tokyo, 2277 Chino Roces Avenue, Legaspi Village, Makati City',
       'Edsa Shangri-La, 1 Garden Way, Ortigas, Mandaluyong City', ...,
       'Kuru�_e��me Mahallesi, Muallim Naci Caddesi, No 56, Be��ikta��, ��stanbul',
       'Kuru�_e��me Mahallesi, Muallim Naci Caddesi, No 64/B, Be��ikta��, ��stanbul',
       'Cafea��a Mahallesi, Bademalt۱ Sokak, No 21/B, Kad۱k�_y, ��stanbul'],
      dtype=object)

In [15]:
df['Address'].value_counts()

Address
Dilli Haat, INA, New Delhi                                                               11
Sector 41, Noida                                                                         11
Greater Kailash (GK) 1, New Delhi                                                        10
The Imperial, Janpath, New Delhi                                                          9
Cyber Hub, DLF Cyber City, Gurgaon                                                        8
                                                                                         ..
23-24, Defence Colony Market, Defence Colony, New Delhi                                   1
28, Main Market, Defence Colony, New Delhi                                                1
Daryaganj, New Delhi                                                                      1
Ground Floor, E-23, Netaji Subhash Marg, Opposite Golcha Cinema, Daryaganj, New Delhi     1
Cafea��a Mahallesi, Bademalt۱ Sokak, No 21/B, Kad۱k�_y, ��stanbul                         1
Name: count, Length: 8918, dtype: int64

In [16]:
df['Locality'].unique()

array(['Century City Mall, Poblacion, Makati City',
       'Little Tokyo, Legaspi Village, Makati City',
       'Edsa Shangri-La, Ortigas, Mandaluyong City', ..., 'Ko��uyolu',
       'Kuru�_e��me', 'Moda'], dtype=object)

In [17]:
df['Locality'].value_counts()

Locality
Connaught Place                                122
Rajouri Garden                                  99
Shahdara                                        87
Defence Colony                                  86
Malviya Nagar                                   85
                                              ... 
Lemon Tree Premier, Sector 29                    1
Omaxe Celebration Mall, Sohna Road, Gurgaon      1
Park Inn, Sector 15, Gurgaon                     1
Plaza Mall, MG Road                              1
Moda                                             1
Name: count, Length: 1208, dtype: int64

In [18]:
df['Cuisines'].unique()

array(['French, Japanese, Desserts', 'Japanese',
       'Seafood, Asian, Filipino, Indian', ..., 'Burger, Izgara',
       'World Cuisine, Patisserie, Cafe', 'Italian, World Cuisine'],
      dtype=object)

In [19]:
df['Cuisines'].value_counts()

Cuisines
North Indian                                             936
North Indian, Chinese                                    511
Chinese                                                  354
Fast Food                                                354
North Indian, Mughlai                                    334
                                                        ... 
Bengali, Fast Food                                         1
North Indian, Rajasthani, Asian                            1
Chinese, Thai, Malaysian, Indonesian                       1
Bakery, Desserts, North Indian, Bengali, South Indian      1
Italian, World Cuisine                                     1
Name: count, Length: 1825, dtype: int64

In [20]:
df['Average Cost for two'].unique()

array([  1100,   1200,   4000,   1500,   1000,   2000,   6000,    800,
          900,    700,    850,    600,   3000,   1800,     55,     30,
           45,     60,     50,     80,     90,    200,    150,    100,
          230,    240,    350,    140,    170,    120,    250,     40,
           70,    300,     85,     75,    400,     65,     25,     10,
            0,     20,     35,      7,    270,    500,    430,    315,
          220,     95,    190,    260,    180,    330,    280,    160,
          285,    130,    110,   2500,   3600,    550,   1400,    450,
          950,    650,   1300,   2400,   1600,   2200,    750,   1700,
         1250,   1900,   2700,   3500,   5000,   2300,   1750,   3200,
         2100,   1850,   4500,   3300,   1350,   1650,   1950,   1050,
          290,   1450,   2800,   3650,   4800,   1550,    720,   6500,
         7000,   1150,   2900,   2600,   4700,   3700,   3800,   4100,
         4300,   2650,   5100,   5500,   4400,   4200,   8000,   2350,
      

In [21]:
df['Average Cost for two'].value_counts()

Average Cost for two
500       900
300       897
400       857
200       687
600       652
         ... 
8000        1
2350        1
150000      1
70000       1
545         1
Name: count, Length: 140, dtype: int64

In [22]:
df['Currency'].unique()

array(['Botswana Pula(P)', 'Brazilian Real(R$)', 'Dollar($)',
       'Emirati Diram(AED)', 'Indian Rupees(Rs.)',
       'Indonesian Rupiah(IDR)', 'NewZealand($)', 'Pounds(��)',
       'Qatari Rial(QR)', 'Rand(R)', 'Sri Lankan Rupee(LKR)',
       'Turkish Lira(TL)'], dtype=object)

In [23]:
df['Currency'].value_counts()

Currency
Indian Rupees(Rs.)        8652
Dollar($)                  482
Pounds(��)                  80
Brazilian Real(R$)          60
Emirati Diram(AED)          60
Rand(R)                     60
NewZealand($)               40
Turkish Lira(TL)            34
Botswana Pula(P)            22
Indonesian Rupiah(IDR)      21
Qatari Rial(QR)             20
Sri Lankan Rupee(LKR)       20
Name: count, dtype: int64

In [24]:
df['Has Table booking'].value_counts()

Has Table booking
No     8393
Yes    1158
Name: count, dtype: int64

In [25]:
df['Has Online delivery'].value_counts()

Has Online delivery
No     7100
Yes    2451
Name: count, dtype: int64

In [26]:
df['Is delivering now'].value_counts()

Is delivering now
No     9517
Yes      34
Name: count, dtype: int64

In [27]:
df['Switch to order menu'].unique()

array(['No'], dtype=object)

In [28]:
df['Switch to order menu'].value_counts()

Switch to order menu
No    9551
Name: count, dtype: int64

In [29]:
df['Price range'].unique()

array([3, 4, 2, 1])

In [30]:
df['Aggregate rating'].value_counts()

Aggregate rating
0.0    2148
3.2     522
3.1     519
3.4     498
3.3     483
3.5     480
3.0     468
3.6     458
3.7     427
3.8     400
2.9     381
3.9     335
2.8     315
4.1     274
4.0     266
2.7     250
4.2     221
2.6     191
4.3     174
4.4     144
2.5     110
4.5      95
2.4      87
4.6      78
4.9      61
2.3      47
4.7      42
2.2      27
4.8      25
2.1      15
2.0       7
1.9       2
1.8       1
Name: count, dtype: int64

In [31]:
df['Rating color'].value_counts()

Rating color
Orange        3737
White         2148
Yellow        2100
Green         1079
Dark Green     301
Red            186
Name: count, dtype: int64

In [32]:
df['Rating text'].value_counts()

Rating text
Average      3737
Not rated    2148
Good         2100
Very Good    1079
Excellent     301
Poor          186
Name: count, dtype: int64

In [33]:
df['Votes'].value_counts()

Votes
0      1094
1       483
2       327
3       244
4       207
       ... 
615       1
538       1
823       1
650       1
901       1
Name: count, Length: 1012, dtype: int64

In [34]:
df = df.drop(['Restaurant Name','Address','Longitude','Latitude','Country Code', 'Locality','Currency','Has Online delivery','Switch to order menu','Rating color','Rating text','Locality Verbose','Is delivering now','Restaurant ID',], axis=1)

In [35]:
df.head(2)

City                    Cuisines  Average Cost for two  \
0  Makati City  French, Japanese, Desserts                  1100   
1  Makati City                    Japanese                  1200   

  Has Table booking  Price range  Aggregate rating  Votes  
0               Yes            3               4.8    314  
1               Yes            3               4.5    591

In [36]:
df['Cuisines'].fillna(df['Cuisines'].mode()[0], inplace=True)

<ipython-input-36-0293b22786c4>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Cuisines'].fillna(df['Cuisines'].mode()[0], inplace=True)


In [37]:
df.isnull().sum()

City                    0
Cuisines                0
Average Cost for two    0
Has Table booking       0
Price range             0
Aggregate rating        0
Votes                   0
dtype: int64


# **modeling**

In [38]:

import pandas as pd
import pickle
from sklearn.preprocessing import LabelEncoder

categorical_columns = [col for col in df.columns if df[col].dtype == 'object']
label_encoders = {}

for col in categorical_columns:
    le = LabelEncoder()
    df[col] = le.fit_transform(df[col])
    label_encoders[col] = le
    with open(f'{col}_encoder.pkl', 'wb') as f:
        pickle.dump(le, f)

In [39]:

x=df.drop(columns={'Cuisines'})
y=df['Cuisines']

**random forest**

In [40]:
from sklearn.model_selection import RandomizedSearchCV, train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score

# Splitting the dataset
X_train, X_temp, y_train, y_temp = train_test_split(x, y, test_size=0.2, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

# Hyperparameter grid for Random Forest
param_grid_rf = {
    'model__n_estimators': [50, 100, 200],
    'model__max_depth': [None, 10, 20, 30],
    'model__min_samples_split': [2, 5, 10],
    'model__min_samples_leaf': [1, 2, 4],
    'model__bootstrap': [True, False]
}

# Random Forest model
random_forest_model = RandomForestClassifier(random_state=42)

# Pipeline
pipeline_rf = Pipeline([
    ('scaler', StandardScaler()),
    ('model', random_forest_model)
])

# RandomizedSearchCV for Random Forest
randomized_search_rf = RandomizedSearchCV(
    pipeline_rf,
    param_distributions=param_grid_rf,
    n_iter=10,
    scoring='accuracy',
    refit=True,
    n_jobs=-1,
    random_state=42,
    verbose=2
)

# Fit the model
print("\nTuning hyperparameters for Random Forest...\n")
randomized_search_rf.fit(X_train, y_train)

# Best estimator
best_rf_model = randomized_search_rf.best_estimator_

# Predictions
y_train_pred_rf = best_rf_model.predict(X_train)
y_val_pred_rf = best_rf_model.predict(X_val)
y_test_pred_rf = best_rf_model.predict(X_test)

# Performance metrics for training data
accuracy_train_rf = accuracy_score(y_train, y_train_pred_rf)
f1_train_rf = f1_score(y_train, y_train_pred_rf, average='weighted')
precision_train_rf = precision_score(y_train, y_train_pred_rf, average='weighted')
recall_train_rf = recall_score(y_train, y_train_pred_rf, average='weighted')

# Performance metrics for validation data
accuracy_val_rf = accuracy_score(y_val, y_val_pred_rf)
f1_val_rf = f1_score(y_val, y_val_pred_rf, average='weighted')
precision_val_rf = precision_score(y_val, y_val_pred_rf, average='weighted')
recall_val_rf = recall_score(y_val, y_val_pred_rf, average='weighted')

# Performance metrics for test data
accuracy_test_rf = accuracy_score(y_test, y_test_pred_rf)
f1_test_rf = f1_score(y_test, y_test_pred_rf, average='weighted')
precision_test_rf = precision_score(y_test, y_test_pred_rf, average='weighted')
recall_test_rf = recall_score(y_test, y_test_pred_rf, average='weighted')

# Print results
print("\nPerformance on Training Data for Random Forest:")
print(f"Accuracy (Train): {accuracy_train_rf:.4f}")
print(f"F1 Score (Train): {f1_train_rf:.4f}")
print(f"Precision (Train): {precision_train_rf:.4f}")
print(f"Recall (Train): {recall_train_rf:.4f}")

print("\nPerformance on Validation Data for Random Forest:")
print(f"Accuracy (Val): {accuracy_val_rf:.4f}")
print(f"F1 Score (Val): {f1_val_rf:.4f}")
print(f"Precision (Val): {precision_val_rf:.4f}")
print(f"Recall (Val): {recall_val_rf:.4f}")

print("\nPerformance on Test Data for Random Forest:")
print(f"Accuracy (Test): {accuracy_test_rf:.4f}")
print(f"F1 Score (Test): {f1_test_rf:.4f}")
print(f"Precision (Test): {precision_test_rf:.4f}")
print(f"Recall (Test): {recall_test_rf:.4f}")


Tuning hyperparameters for Random Forest...

Fitting 5 folds for each of 10 candidates, totalling 50 fits


/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_split.py:805: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py:752: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(



Performance on Training Data for Random Forest:
Accuracy (Train): 0.4305
F1 Score (Train): 0.3915
Precision (Train): 0.5081
Recall (Train): 0.4305

Performance on Validation Data for Random Forest:
Accuracy (Val): 0.1466
F1 Score (Val): 0.0990
Precision (Val): 0.0814
Recall (Val): 0.1466

Performance on Test Data for Random Forest:
Accuracy (Test): 0.1224
F1 Score (Test): 0.0766
Precision (Test): 0.0655
Recall (Test): 0.1224


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_

**lbgm**

In [41]:
from sklearn.model_selection import RandomizedSearchCV, train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from lightgbm import LGBMClassifier
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score

# Splitting the dataset
X_train, X_temp, y_train, y_temp = train_test_split(x, y, test_size=0.2, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

# Hyperparameter grid for LightGBM
param_grid_lgbm = {
    'model__num_leaves': [31, 50, 100],
    'model__learning_rate': [0.01, 0.1, 0.2],
    'model__n_estimators': [50, 100, 200],
    'model__max_depth': [-1, 10, 20]
}

# LightGBM model
lgbm_model = LGBMClassifier(random_state=42)

# Pipeline
pipeline_lgbm = Pipeline([
    ('scaler', StandardScaler()),
    ('model', lgbm_model)
])

# RandomizedSearchCV for LightGBM
randomized_search_lgbm = RandomizedSearchCV(
    pipeline_lgbm,
    param_distributions=param_grid_lgbm,
    n_iter=10,
    scoring='accuracy',
    refit=True,
    n_jobs=-1,
    random_state=42,
    verbose=2
)

# Fit the model
print("\nTuning hyperparameters for LightGBM...\n")
randomized_search_lgbm.fit(X_train, y_train)

# Best estimator
best_lgbm_model = randomized_search_lgbm.best_estimator_

# Predictions
y_train_pred_lgbm = best_lgbm_model.predict(X_train)
y_val_pred_lgbm = best_lgbm_model.predict(X_val)
y_test_pred_lgbm = best_lgbm_model.predict(X_test)

# Performance metrics for training data
accuracy_train_lgbm = accuracy_score(y_train, y_train_pred_lgbm)
f1_train_lgbm = f1_score(y_train, y_train_pred_lgbm, average='weighted')
precision_train_lgbm = precision_score(y_train, y_train_pred_lgbm, average='weighted')
recall_train_lgbm = recall_score(y_train, y_train_pred_lgbm, average='weighted')

# Performance metrics for validation data
accuracy_val_lgbm = accuracy_score(y_val, y_val_pred_lgbm)
f1_val_lgbm = f1_score(y_val, y_val_pred_lgbm, average='weighted')
precision_val_lgbm = precision_score(y_val, y_val_pred_lgbm, average='weighted')
recall_val_lgbm = recall_score(y_val, y_val_pred_lgbm, average='weighted')

# Performance metrics for test data
accuracy_test_lgbm = accuracy_score(y_test, y_test_pred_lgbm)
f1_test_lgbm = f1_score(y_test, y_test_pred_lgbm, average='weighted')
precision_test_lgbm = precision_score(y_test, y_test_pred_lgbm, average='weighted')
recall_test_lgbm = recall_score(y_test, y_test_pred_lgbm, average='weighted')

# Print results
print("\nPerformance on Training Data for LightGBM:")
print(f"Accuracy (Train): {accuracy_train_lgbm:.4f}")
print(f"F1 Score (Train): {f1_train_lgbm:.4f}")
print(f"Precision (Train): {precision_train_lgbm:.4f}")
print(f"Recall (Train): {recall_train_lgbm:.4f}")

print("\nPerformance on Validation Data for LightGBM:")
print(f"Accuracy (Val): {accuracy_val_lgbm:.4f}")
print(f"F1 Score (Val): {f1_val_lgbm:.4f}")
print(f"Precision (Val): {precision_val_lgbm:.4f}")
print(f"Recall (Val): {recall_val_lgbm:.4f}")

print("\nPerformance on Test Data for LightGBM:")
print(f"Accuracy (Test): {accuracy_test_lgbm:.4f}")
print(f"F1 Score (Test): {f1_test_lgbm:.4f}")
print(f"Precision (Test): {precision_test_lgbm:.4f}")
print(f"Recall (Test): {recall_test_lgbm:.4f}")


/usr/local/lib/python3.10/dist-packages/dask/dataframe/__init__.py:42: FutureWarning: 
Dask dataframe query planning is disabled because dask-expr is not installed.

You can install it with `pip install dask[dataframe]` or `conda install dask`.
This will raise in a future version.

  warnings.warn(msg, FutureWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/utils/_tags.py:354: FutureWarning: The LGBMClassifier or classes from which it inherits use `_get_tags` and `_more_tags`. Please define the `__sklearn_tags__` method, or inherit from `sklearn.base.BaseEstimator` and/or other appropriate mixins such as `sklearn.base.TransformerMixin`, `sklearn.base.ClassifierMixin`, `sklearn.base.RegressorMixin`, and `sklearn.base.OutlierMixin`. From scikit-learn 1.7, not defining `__sklearn_tags__` will raise an error.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_split.py:805: UserWarning: The least populated class in y has only 1 members, which is less th


Tuning hyperparameters for LightGBM...

Fitting 5 folds for each of 10 candidates, totalling 50 fits


/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py:752: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.001142 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 470
[LightGBM] [Info] Number of data points in the train set: 7640, number of used features: 6
[LightGBM] [Info] Start training from score -7.554859
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -5.683056
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [Info] Start training from score -8.941153
[LightGBM] [

/usr/local/lib/python3.10/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/utils/_tags.py:354: FutureWarning: The LGBMClassifier or classes from which it inherits use `_get_tags` and `_more_tags`. Please define the `__sklearn_tags__` method, or inherit from `sklearn.base.BaseEstimator` and/or other appropriate mixins such as `sklearn.base.TransformerMixin`, `sklearn.base.ClassifierMixin`, `sklearn.base.RegressorMixin`, and `sklearn.base.OutlierMixin`. From scikit-learn 1.7, not defining `__sklearn_tags__` will raise an error.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/utils/_tags.py:354: FutureWarning: The L


Performance on Training Data for LightGBM:
Accuracy (Train): 0.5508
F1 Score (Train): 0.5405
Precision (Train): 0.7054
Recall (Train): 0.5508

Performance on Validation Data for LightGBM:
Accuracy (Val): 0.1445
F1 Score (Val): 0.0862
Precision (Val): 0.0889
Recall (Val): 0.1445

Performance on Test Data for LightGBM:
Accuracy (Test): 0.1287
F1 Score (Test): 0.0751
Precision (Test): 0.0701
Recall (Test): 0.1287


/usr/local/lib/python3.10/dist-packages/sklearn/utils/_tags.py:354: FutureWarning: The LGBMClassifier or classes from which it inherits use `_get_tags` and `_more_tags`. Please define the `__sklearn_tags__` method, or inherit from `sklearn.base.BaseEstimator` and/or other appropriate mixins such as `sklearn.base.TransformerMixin`, `sklearn.base.ClassifierMixin`, `sklearn.base.RegressorMixin`, and `sklearn.base.OutlierMixin`. From scikit-learn 1.7, not defining `__sklearn_tags__` will raise an error.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels 

**decision tree**


In [42]:
from sklearn.model_selection import RandomizedSearchCV, train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score

# Splitting the dataset
X_train, X_temp, y_train, y_temp = train_test_split(x, y, test_size=0.2, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

# Hyperparameter grid for Decision Tree
param_grid_dt = {
    'model__criterion': ['gini', 'entropy'],  # Splitting criteria
    'model__max_depth': [None, 10, 20, 30],  # Maximum depth of the tree
    'model__min_samples_split': [2, 5, 10],  # Minimum samples required to split a node
    'model__min_samples_leaf': [1, 2, 4]     # Minimum samples required in a leaf node
}

# Decision Tree model
decision_tree_model = DecisionTreeClassifier(random_state=42)

# Pipeline
pipeline_dt = Pipeline([
    ('scaler', StandardScaler()),
    ('model', decision_tree_model)
])

# RandomizedSearchCV for Decision Tree
randomized_search_dt = RandomizedSearchCV(
    pipeline_dt,
    param_distributions=param_grid_dt,
    n_iter=10,
    scoring='accuracy',
    refit=True,
    n_jobs=-1,
    random_state=42,
    verbose=2
)

# Fit the model
print("\nTuning hyperparameters for Decision Tree...\n")
randomized_search_dt.fit(X_train, y_train)

# Best estimator
best_dt_model = randomized_search_dt.best_estimator_

# Predictions
y_train_pred_dt = best_dt_model.predict(X_train)
y_val_pred_dt = best_dt_model.predict(X_val)
y_test_pred_dt = best_dt_model.predict(X_test)

# Performance metrics for training data
accuracy_train_dt = accuracy_score(y_train, y_train_pred_dt)
f1_train_dt = f1_score(y_train, y_train_pred_dt, average='weighted')
precision_train_dt = precision_score(y_train, y_train_pred_dt, average='weighted')
recall_train_dt = recall_score(y_train, y_train_pred_dt, average='weighted')

# Performance metrics for validation data
accuracy_val_dt = accuracy_score(y_val, y_val_pred_dt)
f1_val_dt = f1_score(y_val, y_val_pred_dt, average='weighted')
precision_val_dt = precision_score(y_val, y_val_pred_dt, average='weighted')
recall_val_dt = recall_score(y_val, y_val_pred_dt, average='weighted')

# Performance metrics for test data
accuracy_test_dt = accuracy_score(y_test, y_test_pred_dt)
f1_test_dt = f1_score(y_test, y_test_pred_dt, average='weighted')
precision_test_dt = precision_score(y_test, y_test_pred_dt, average='weighted')
recall_test_dt = recall_score(y_test, y_test_pred_dt, average='weighted')

# Print results
print("\nPerformance on Training Data for Decision Tree:")
print(f"Accuracy (Train): {accuracy_train_dt:.4f}")
print(f"F1 Score (Train): {f1_train_dt:.4f}")
print(f"Precision (Train): {precision_train_dt:.4f}")
print(f"Recall (Train): {recall_train_dt:.4f}")

print("\nPerformance on Validation Data for Decision Tree:")
print(f"Accuracy (Val): {accuracy_val_dt:.4f}")
print(f"F1 Score (Val): {f1_val_dt:.4f}")
print(f"Precision (Val): {precision_val_dt:.4f}")
print(f"Recall (Val): {recall_val_dt:.4f}")

print("\nPerformance on Test Data for Decision Tree:")
print(f"Accuracy (Test): {accuracy_test_dt:.4f}")
print(f"F1 Score (Test): {f1_test_dt:.4f}")
print(f"Precision (Test): {precision_test_dt:.4f}")
print(f"Recall (Test): {recall_test_dt:.4f}")



Tuning hyperparameters for Decision Tree...

Fitting 5 folds for each of 10 candidates, totalling 50 fits


/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_split.py:805: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(



Performance on Training Data for Decision Tree:
Accuracy (Train): 0.2435
F1 Score (Train): 0.1877
Precision (Train): 0.2160
Recall (Train): 0.2435

Performance on Validation Data for Decision Tree:
Accuracy (Val): 0.1141
F1 Score (Val): 0.0832
Precision (Val): 0.0716
Recall (Val): 0.1141

Performance on Test Data for Decision Tree:
Accuracy (Test): 0.1067
F1 Score (Test): 0.0740
Precision (Test): 0.0613
Recall (Test): 0.1067


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_

**Performance on Test Data for Random Forest:**          
Performance on Test Data for Random Forest:             
Accuracy (Test): 0.1224                         
F1 Score (Test): 0.0766                                 
Precision (Test): 0.0655                         
Recall (Test): 0.1224                    
**Performance on Test Data for LightGBM:**                       
Accuracy (Test): 0.1287                    
F1 Score (Test): 0.0751                                 
Precision (Test): 0.0701                                   
Recall (Test): 0.1287                                      
**Performance on Test Data for Decision Tree:**                    
Performance on Test Data for Decision Tree:              
Accuracy (Test): 0.1067                  
F1 Score (Test): 0.0740               
Precision (Test): 0.0613                      
Recall (Test): 0.1067                          





**Summary of Model Performance:**                
**LightGBM:**          
LightGBM achieved the highest test accuracy (0.1287) and recall (1287) among the three models.This suggests that LightGBM has a slight edge in identifying positive cases correctly, even though the overall performance is still limited             

**Random Forest:**           
Random Forest came second with an accuracy of 0.1224 and recall of 01224.Its precision (0.0655) and F1 score (0.0766) are slightly higher than the Decision Tree but lower than LightGBM, indicating it is moderately effective.           

**Decision Tree:**            
The Decision Tree had the lowest accuracy (0.1067) and the weakest
performance across all metrics, including precision (0.0613) and F1 score (0.0740).This implies it is not well-suited for the dataset in its current state.


Based on the above metrics, LightGBM  is the preferred model for deployment, given its slight superiority.

I am encountering several challenges in my work. Despite trying various models such as XGBoost, LightGBM, Random Forest, Decision Tree, Logistic Regression, and SVM, only Random Forest, XGBoost, and Decision Tree showed some success. However, these models are still underfitting, as both training and testing accuracies are poor. I have attempted parameter tuning and applied regularization, but the results remain largely unchanged or show only slight improvements. I also implemented cross-validation and tried making the data more complex, but the evaluation metrics still show no significant improvement.